teleport proxy crash

## Description

Teleport proxy can sometimes crash. 

Versions affected:

Teleport >= 2.2.3

Relevant lines:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xd45193]
goroutine 1501 [running]:
github.com/gravitational/teleport/lib/srv.(*proxySubsys).proxyToSite(0xc4213ce070, 0x0, 0x0, 0x1845340, 0xc42132e390, 0x184d800, 0xc4207b2000, 0x7fae03456c20, 0xc4213ce070)
    /gopath/src/github.com/gravitational/teleport/lib/srv/proxy.go:171 +0x63
github.com/gravitational/teleport/lib/srv.(*proxySubsys).start(0xc4213ce070, 0xc421f02360, 0x184d800, 0xc4207b2000, 0xc42128e800, 0xc4219da0e0, 0x7fae03456c20, 0xc422202c58)
    /gopath/src/github.com/gravitational/teleport/lib/srv/proxy.go:159 +0x4a8
github.com/gravitational/teleport/lib/srv.(*Server).handleSubsystem(0xc4205c0e00, 0x184d800, 0xc4207b2000, 0xc42128e800, 0xc4219da0e0, 0x2, 0xc42006f400)
    /gopath/src/github.com/gravitational/teleport/lib/srv/sshserver.go:1049 +0x24f
github.com/gravitational/teleport/lib/srv.(*Server).dispatch(0xc4205c0e00, 0x184d800, 0xc4207b2000, 0xc42128e800, 0xc4219da0e0, 0xc421fa4e20, 0x28)
    /gopath/src/github.com/gravitational/teleport/lib/srv/sshserver.go:915 +0x2bc
github.com/gravitational/teleport/lib/srv.(*Server).handleSessionRequests(0xc4205c0e00, 0xc421f02360, 0x184d800, 0xc4207b2000, 0xc4219b4000)
    /gopath/src/github.com/gravitational/teleport/lib/srv/sshserver.go:886 +0x2e5
created by github.com/gravitational/teleport/lib/srv.(*Server).HandleNewChan
    /gopath/src/github.com/gravitational/teleport/lib/srv/sshserver.go:722 +0x29c
```

--- a/lib/srv/proxy.go
+++ b/lib/srv/proxy.go
@@ -95,7 +95,7 @@ func parseProxySubsys(request string, srv *Server) (*proxySubsys, error) {
 		if clusterName == "" && targetHost == "" {
 			return nil, trace.BadParameter("invalid format for proxy request: missing cluster name or target host in %q", request)
 		}
-	case len(parts) > 3: // "proxy:host:22@namespace@clustername"
+	case len(parts) >= 3: // "proxy:host:22@namespace@clustername"
 		clusterName = strings.Join(parts[2:], "@")
 		namespace = parts[1]
 		targetHost, targetPort, err = utils.SplitHostPort(parts[0])

--- a/lib/srv/proxy.go
+++ b/lib/srv/proxy.go
@@ -59,56 +59,68 @@ type proxySubsys struct {
 //  "proxy:@clustername"        - Teleport request to connect to an auth server for cluster with name 'clustername'
 //  "proxy:host:22@clustername" - Teleport request to connect to host:22 on cluster 'clustername'
 //  "proxy:host:22@namespace@clustername"
-func parseProxySubsys(name string, srv *Server) (*proxySubsys, error) {
-	log.Debugf("parse_proxy_subsys(%s)", name)
+func parseProxySubsys(request string, srv *Server) (*proxySubsys, error) {
+	log.Debugf("parse_proxy_subsys(%q)", request)
 	var (
-		clusterName string
-		host        string
-		port        string
-		paramError  = trace.BadParameter("invalid format for proxy request: '%v', expected 'proxy:host:port@site'", name)
+		clusterName  string
+		targetHost   string
+		targetPort   string
+		paramMessage = fmt.Sprintf("invalid format for proxy request: %q, expected 'proxy:host:port@cluster'", request)
 	)
 	const prefix = "proxy:"
 	// get rid of 'proxy:' prefix:
-	if strings.Index(name, prefix) != 0 {
-		return nil, trace.Wrap(paramError)
+	if strings.Index(request, prefix) != 0 {
+		return nil, trace.BadParameter(paramMessage)
 	}
-	name = strings.TrimPrefix(name, prefix)
+	requestBody := strings.TrimPrefix(request, prefix)
 	namespace := defaults.Namespace
-	// find the site name in the argument:
-	parts := strings.Split(name, "@")
-	switch len(parts) {
-	case 2:
-		clusterName = strings.Join(parts[1:], "@")
-		name = parts[0]
-	case 3:
+	var err error
+	parts := strings.Split(requestBody, "@")
+	switch {
+	case len(parts) == 0: // "proxy:"
+		return nil, trace.BadParameter(paramMessage)
+	case len(parts) == 1: // "proxy:host:22"
+		targetHost, targetPort, err = utils.SplitHostPort(parts[0])
+		if err != nil {
+			return nil, trace.BadParameter(paramMessage)
+		}
+	case len(parts) == 2: // "proxy:@clustername" or "proxy:host:22@clustername"
+		if parts[0] != "" {
+			targetHost, targetPort, err = utils.SplitHostPort(parts[0])
+			if err != nil {
+				return nil, trace.BadParameter(paramMessage)
+			}
+		}
+		clusterName = parts[1]
+		if clusterName == "" && targetHost == "" {
+			return nil, trace.BadParameter("invalid format for proxy request: missing cluster name or target host in %q", request)
+		}
+	case len(parts) > 3: // "proxy:host:22@namespace@clustername"
 		clusterName = strings.Join(parts[2:], "@")
 		namespace = parts[1]
-		name = parts[0]
-	}
-	// find host & port in the arguments:
-	host, port, err := net.SplitHostPort(name)
-	if clusterName == "" && err != nil {
-		return nil, trace.Wrap(paramError)
+		targetHost, targetPort, err = utils.SplitHostPort(parts[0])
+		if err != nil {
+			return nil, trace.BadParameter(paramMessage)
+		}
 	}
 	if clusterName != "" && srv.proxyTun != nil {
 		_, err := srv.proxyTun.GetSite(clusterName)
 		if err != nil {
-			return nil, trace.BadParameter("unknown cluster '%s'", clusterName)
+			return nil, trace.BadParameter("invalid format for proxy request: unknown cluster %q in %q", clusterName, request)
 		}
 	}
-
 	return &proxySubsys{
 		namespace: namespace,
 		srv:       srv,
-		host:      host,
-		port:      port,
+		host:      targetHost,
+		port:      targetPort,
 		siteName:  clusterName,
 		closeC:    make(chan struct{}),
 	}, nil
 }
 
 func (t *proxySubsys) String() string {
-	return fmt.Sprintf("proxySubsys(site=%s/%s, host=%s, port=%s)",
+	return fmt.Sprintf("proxySubsys(cluster=%s/%s, host=%s, port=%s)",
 		t.namespace, t.siteName, t.host, t.port)
 }
 

--- a/lib/srv/proxy_test.go
+++ b/lib/srv/proxy_test.go
@@ -68,3 +68,21 @@ func (s *ProxyTestSuite) TestParseProxyRequest(c *check.C) {
 	c.Assert(subsys.port, check.Equals, "100")
 	c.Assert(subsys.siteName, check.Equals, "moon")
 }
+
+func (s *ProxyTestSuite) TestParseBadRequests(c *check.C) {
+	testCases := []string{
+		// empty request
+		"proxy:",
+		// missing hostname
+		"proxy::80",
+		// missing hostname and missing cluster name
+		"proxy:@",
+		// just random string
+		"this is bad string",
+	}
+	for _, input := range testCases {
+		comment := check.Commentf("test case: %q", input)
+		_, err := parseProxySubsys(input, s.srv)
+		c.Assert(err, check.NotNil, comment)
+	}
+}

--- a/lib/srv/proxy_test.go
+++ b/lib/srv/proxy_test.go
@@ -67,6 +67,16 @@ func (s *ProxyTestSuite) TestParseProxyRequest(c *check.C) {
 	c.Assert(subsys.host, check.Equals, "station")
 	c.Assert(subsys.port, check.Equals, "100")
 	c.Assert(subsys.siteName, check.Equals, "moon")
+
+	// proxy request for the host:port@namespace@cluster
+	subsys, err = parseProxySubsys("proxy:station:100@system@moon", s.srv)
+	c.Assert(err, check.IsNil)
+	c.Assert(subsys, check.NotNil)
+	c.Assert(subsys.srv, check.Equals, s.srv)
+	c.Assert(subsys.host, check.Equals, "station")
+	c.Assert(subsys.port, check.Equals, "100")
+	c.Assert(subsys.siteName, check.Equals, "moon")
+	c.Assert(subsys.namespace, check.Equals, "system")
 }
 
 func (s *ProxyTestSuite) TestParseBadRequests(c *check.C) {

--- a/lib/utils/utils.go
+++ b/lib/utils/utils.go
@@ -33,6 +33,18 @@ import (
 	"golang.org/x/crypto/ssh"
 )
 
+// SplitHostPort splits host and port and checks that host is not empty
+func SplitHostPort(hostname string) (string, string, error) {
+	host, port, err := net.SplitHostPort(hostname)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	if host == "" {
+		return "", "", trace.BadParameter("empty hostname")
+	}
+	return host, port, nil
+}
+
 type HostKeyCallback func(hostID string, remote net.Addr, key ssh.PublicKey) error
 
 func ReadPath(path string) ([]byte, error) {

